wdl/src/main/scala/wdl/command/ParameterCommandPart.scala

```diff
@@ -49,29 +45,50 @@ case class ParameterCommandPart(attributes: Map[String, String], expression: Wdl
+    // Note that evaluating this expression may have the side effect of causing a file to be created if `writeFile`
+    // is invoked. That file will be written to an "engine-relative" location which may be in cloud storage.
+    val evaluatedCommandPartExpression: ErrorOr[WomValue] = expression.evaluate(lookup, functions) match {
       case Success(v) => v match {
-        case WomOptionalValue(_, opt) => opt.getOrElse(defaultString)
-        case _ => v
+        case WomOptionalValue(_, opt) => opt.getOrElse(defaultString).validNel
+        case _ => v.validNel
       }
-      case Failure(OptionalNotSuppliedException(_)) => defaultString
-      case Failure(f) => throw new UnsupportedOperationException(s"Could not evaluate expression: ${expression.toWomString}", f)
+      case Failure(OptionalNotSuppliedException(_)) => defaultString.validNel
+      case Failure(_) => s"Could not evaluate expression: ${expression.toWomString}".invalidNel
     }
 
-    valueMapper(value) match {
-      case b: WomBoolean if attributes.contains("true") && attributes.contains("false") => if (b.value) attributes.get("true").head else attributes.get("false").head
-      case p: WomPrimitive => p.valueString
-      case a: WomArray if attributes.contains("sep") => a.value.map(_.valueString).mkString(attributes.get("sep").head)
-      case _: WomArray => throw new UnsupportedOperationException(s"Expression '${expression.toString}' evaluated to an Array but no 'sep' was specified")
-      case _ => throw new UnsupportedOperationException(s"Could not string-ify value: $value")
+    // Do the two arguments represent the same WomFile path.
+    def representTheSamePath(womFile: WomFile, maybeWomFile: WomValue): Boolean = {
+      Option(maybeWomFile) collect { case isWomFile: WomFile if isWomFile.value == womFile.value => true } isDefined
     }
-  }
 
-  private def defaultString = {
-    if (attributes.contains("default")) {
-      WomString(attributes.get("default").head)
-    } else {
-      WomString("")
+    // Create the stringified version of the command and record any file created by `writeFile`.
+    def instantiateCommand(value: WomValue): ErrorOr[InstantiatedCommand] = {
+      valueMapper(value) match {
+        case b: WomBoolean if attributes.contains("true") && attributes.contains("false") =>
+          InstantiatedCommand(if (b.value) attributes.get("true").head else attributes.get("false").head).validNel
+        case f: WomFile if !representTheSamePath(f, value) =>
+          // Files generated by writeFiles have "engine-relative" paths which will be different from the container paths
+          // calculated by `valueMapper`. "engine-relative" may mean either the non-Docker container path on the host
+          // running Cromwell, or a cloud path. Capture these newly created files and their engine paths here.
+          val enginePathFile = value.asInstanceOf[WomFile]
```

Comment:
ToL: this assumes that the `valueMapper` maps only `WomFile`s to `WomFile`s right ?
Since it's provided by the backend and could potentially be anything we might want to safe guard against weird value mappers here that would map `WomSomethingElse` to `WomFile`. Or at least leave a comment.

Reply:
Yeah this isn't quite right. I think it was right at one point but I refactored it to be not right. Good catch!